Updated runPromptReco.py

andresfelquintero opened the pull request:
#### PR description: This update was based on what was discussed on the following JIRA ticket https://its.cern.ch/jira/browse/CMSTZ-427 

<!-- runPromptReco.py in test folder was updated, because it did not take the number of cores/threads into consideration to create configuration files. -->

#### PR validation:

<!-- No new unit tests were created, it was tested on a local environment and will complete a wrapper-->

#### if this PR is a backport please specify the original PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

andresfelquintero committed:
Updated runPromptReco.py

I made the change suggested by Slava.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26684/9656

- This PR adds an extra 12KB to repository

cmsbuild commented:
A new Pull Request was created by  @andresfelquintero  for master.

It involves the following packages:

Configuration/DataProcessing


@cmsbuild, @franzoni, @fabiocos, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
@andresfelquintero 
please add a link to the T0 JIRA ticket in the PR description for better self-documentation

andresfelquintero commented:
> @andresfelquintero
> please add a link to the T0 JIRA ticket in the PR description for better self-documentation

It was added to the description as suggested.

cmsbuild commented:
+1
Tested at: 9756b20ebe88c9a8c4a6ffb1ec01282bbcace35f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7aac08/60/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/bb02d607c4e82bbcd23da0f87b3dcbc40d0d6087
https://github.com/cms-sw/cmssw/commit/c458253940db67f994cf13c16cbebbc53faa44be
https://github.com/cms-sw/cmssw/commit/3fa9341a93f7e108800e84fc00d583ac2550e308
https://github.com/cms-sw/cmssw/commit/e848c05da8792f89b5ac2886f91a40a95ef60b7e
https://github.com/cms-sw/cmssw/commit/687a920aae438d7997949535772d523ae995309c
https://github.com/cms-sw/cmssw/commit/e0d7bdb96ec5ce69429b36d23bf30dbbafccba6b
https://github.com/cms-sw/cmssw/commit/8ce71450a51042e4421cbe9fcf720da3be00329c
https://github.com/cms-sw/cmssw/commit/f686c846bba3b2fb55594d5a5b21ead854400914
https://github.com/cms-sw/cmssw/commit/14d40ab34218f12a21ca27f2beba3b86d9043aee
https://github.com/cms-sw/cmssw/commit/707d82de5348e8aa604b992e9b040f53f3e0e61f
https://github.com/cms-sw/cmssw/commit/aec25f3371ee30f8af01e4dc5aec2921674a0a78
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26684/60/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26684/60/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7aac08/60/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3211964
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211759
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

davidlange6 commented:
Two comments from my side
- nThreads would be a more consistent name for this option
- a default of 1 is likely better than requiring this option to be set (that way, the examples in the help will not be broken by this PR)

fabiocos commented:
@smuzaffar interesting situation, the bot tells us that the PR was updated but I see just one commit and no message about force updates

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

fabiocos commented:
please test

smuzaffar commented:
@fabiocos , looks like github timestamps are messed up. As you can see comments from @davidlange6 and @slava77 are also in future. 
The message from @cmsbuild is also in future but if I see the jenkins log the message was added at

```
The following comment will be made:

Latest commit sha:  9756b20ebe88c9a8c4a6ffb1ec01282bbcace35f
PR update time 2019-05-08 11:48:23
Time UTC: 2019-05-08 11:48:49.274511
```
which is correct timestamp (and matches the jenkins job's time).

andresfelquintero commented:
> Two comments from my side
> 
> * nThreads would be a more consistent name for this option
> * a default of 1 is likely better than requiring this option to be set (that way, the examples in the help will not be broken by this PR)

So @slava77 , should I change the name of it? For us it is better to use the name numCores because it is the same name of the variable that we obtain from the API, but if this is more useful for you then I will change the name as requested by @davidlange6 .

smuzaffar commented:
abort

cmsbuild commented:
Jenkins tests are aborted.

smuzaffar commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

smuzaffar commented:
@slava77 and @davidlange6 can you please add any comment here? looks like cmsbot timestamp is correct now. I would like to confirm that your messages are also in present or not?

fabiocos commented:
Should we retrigger this test? Now let's see whether I can read this comment

fabiocos commented:
@smuzaffar the problem seems solved in the sense that I see the message at a reasonable time, but the messages that were moved in the future are still seen in the future, this probably will stay right?

davidlange6 commented:
here is a comment

> On May 8, 2019, at 6:13 PM, Malik Shahzad Muzaffar <notifications@github.com> wrote:
> 
> @slava77 and @davidlange6 can you please add any comment here? looks like cmsbot timestamp is correct now. I would like to confirm that your messages are also in present or not?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
> So @slava77 , should I change the name of it? For us it is better to use the name numCores because it is the same name of the variable that we obtain from the API, but if this is more useful for you then I will change the name as requested by @davidlange6 .

For my reference, please provide a pointer to the T0/WMAgent code where this is used as "numCores".
Thank you.

slava77 commented:
On 5/8/19 9:17 AM, Fabio Cossutti wrote:
> the problem seems solved in the sense that I see the message at a
> reasonable time, but the messages that were moved in the future are
> still seen in the future, this probably will stay right?
> 

my messages are however sometimes moved into the past (at least in my
rendering/display).
The most recent ones are earlier than what I've sent properly earlier.
I'm trying to rely on what I get in my inbox instead of the web right
now to read the messages.

slava77 commented:
this is what's showing up in my browser (taking the last few)

<img width="648" alt="Screen Shot 2019-05-08 at 10 04 24 AM" src="https://user-images.githubusercontent.com/4676718/57393791-b64ca000-7178-11e9-8f1f-5db27e5437f6.png">

cmsbuild commented:
+1
Tested at: 9756b20ebe88c9a8c4a6ffb1ec01282bbcace35f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7aac08/89/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/bb02d607c4e82bbcd23da0f87b3dcbc40d0d6087
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26684/89/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26684/89/git-merge-result

cmsbuild commented:
Comparison job queued.

andresfelquintero commented:
> > So @slava77 , should I change the name of it? For us it is better to use the name numCores because it is the same name of the variable that we obtain from the API, but if this is more useful for you then I will change the name as requested by @davidlange6 .
> 
> For my reference, please provide a pointer to the T0/WMAgent code where this is used as "numCores".
> Thank you.

Hello. You were right, there was no code set for that name, I just set that name for my wrapper based on the outcome of this service. https://cmsweb.cern.ch/t0wmadatasvc/prod/reco_config?run=326607 So I will change the name as @davidlange6 suggested.

andresfelquintero committed:
Updated RunPromptReco.py on test folder

I changed the name of numCores variable to nThreads as suggested by David Lange

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26684/9690

- This PR adds an extra 16KB to repository

andresfelquintero commented:
Is it ok for this pull request has two commits? or should I comprise them into a single one? I don't know how you handle them.

slava77 commented:
> > > So @slava77 , should I change the name of it? For us it is better to use the name numCores because it is the same name of the variable that we obtain from the API, but if this is more useful for you then I will change the name as requested by @davidlange6 .
> > 
> > 
> > For my reference, please provide a pointer to the T0/WMAgent code where this is used as "numCores".
> > Thank you.
> 
> Hello. You were right, there was no code set for that name, I just set that name for my wrapper based on the outcome of this service. https://cmsweb.cern.ch/t0wmadatasvc/prod/reco_config?run=326607 So I will change the name as @davidlange6 suggested.

but even here the parameter is apparently called "multicore", not "numCores".



> Is it ok for this pull request has two commits? or should I comprise them into a single one? I don't know how you handle them.

either way is OK.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7aac08/89/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3211964
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211759
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

cmsbuild commented:
+1
Tested at: d9e09762ad579777ebfdfa2fdae2778e1fac75c9
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7aac08/97/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7aac08/97/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3211964
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211759
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

cmsbuild commented:
Pull request #26684 was updated. @franzoni, @fabiocos, @davidlange6 can you please check and sign again.

andresfelquintero commented:
So the checks have passed, do you have any other comments? @slava77 @davidlange6

fabiocos commented on Configuration/DataProcessing/test/RunPromptReco.py:
```diff
@@ -45,6 +43,8 @@ def __call__(self):
             msg = "No --lfn specified"
             raise RuntimeError(msg)
+        if self.nThreads == None:
+            msg = "No --nThreads specified"
```
@andresfelquintero I do not see the purpose of this statement, if no action or printout is added, the "msg" string is not used as it is now

fabiocos replied:
@andresfelquintero wait, in this case the comment of @davidlange6 is valid, you cannot have ```None``` as default value, but it should be 1, as in https://cmssdt.cern.ch/lxr/source/Configuration/Applications/python/ConfigBuilder.py#0087
otherwise you always force the user to specify explicitly the option

andresfelquintero replied:
so @slava77 do you think it is better to have a default value? or shall we let that input as a completely optional one?

slava77 replied:
in configurations built with cmsDriver the parameter is optional

andresfelquintero replied:
If this is going to be an optional input for this code, then there is no point in having the error message or the if statement on line 46. Shall I delete it then?

fabiocos commented on Configuration/DataProcessing/test/RunPromptReco.py:
```diff
@@ -196,6 +195,8 @@ def __call__(self):
         if opt == "--no-output":
             recoinator.noOutput = True
+        if opt == "--nThreads":
+            recoinator.nThreads = arg
```
@andresfelquintero passing the option is not enough, it must be used within recoinator (which is not the case now)

fabiocos commented:
@andresfelquintero @slava77 @davidlange6 as far as I can see this PR is so far dummy, as in https://cmssdt.cern.ch/lxr/source/Configuration/DataProcessing/python/Reco.py#0045 the passed argument is used nowhere now. Adding a statement like:

```
        if 'nThreads' in args:
            options.nThreads=args['nThreads']
```

passes the argument to ConfigBuilder. Still when I have tried this I am puzzled by the fact that the option seems to disappear, even if in https://cmssdt.cern.ch/lxr/source/Configuration/Applications/python/ConfigBuilder.py#2224 the lines are properly added (verified by printing the content of self.pythonCfgCode), and this is still true at the end of the ConfigBuilder prepare function

fabiocos commented:
ok, in cmsDriver the pythonCfgCode is dumped into a file, if I do the same within Reco.py I get the proper configuration dumped. But if I print process.options the numberOfThreads option is just not there

slava77 commented:
> @andresfelquintero @slava77 @davidlange6 as far as I can see this PR is so far dummy, as in https://cmssdt.cern.ch/lxr/source/Configuration/DataProcessing/python/Reco.py#0045 the passed argument is used nowhere now. Adding a statement like:

you are right.
I confused myself with some other options "magically" appearing in the ConfigBuilder args list.
I think that the Reco.py should be updated.

andresfelquintero committed:
Updated RunPromptReco.py

I added an error message when the argument nthreads is not passed.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26684/9732

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #26684 was updated. @cmsbuild, @franzoni, @fabiocos, @davidlange6 can you please check and sign again.

andresfelquintero committed:
Update Reco.py

Added an if statement suggested by Fabio Cossutti.

cmsbuild commented:
The code-checks are being triggered in jenkins.

andresfelquintero commented:
I just edited Reco.py as suggested and also created the error message on RunPromptReco.py. Thanks for the feedback.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26684/9738

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #26684 was updated. @cmsbuild, @franzoni, @fabiocos, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

fabiocos commented:
@andresfelquintero general comment: the express case should be treated in the same way as prompt reco, in general we are having even more DQM troubles in express than in prompt...

andresfelquintero commented:
> @andresfelquintero general comment: the express case should be treated in the same way as prompt reco, in general we are having even more DQM troubles in express than in prompt...

That is correct, shall I change Express as well on this pull request or should I open another one? I was just changing this file because my wrapper is working for PromptReco only, but I will create another one for Express as soon as we confirm that everything works perfectly with the first wrapper.

cmsbuild commented:
-1

Tested at: d1d74add68ad0f8a5e33da1db7c4592126c5a816

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7aac08/153/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test TestConfigDP had ERRORS

cmsbuild commented:
Comparison job queued.

andresfelquintero committed:
Updated PromptReco.py to make nthreads optional

I deleted the lines that checked for a runtime error if nthreads was None.

andresfelquintero committed:
Merge pull request #1 from andresfelquintero/recoWrapper1

Updated PromptReco.py to make nthreads optional

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26684/9740

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #26684 was updated. @cmsbuild, @franzoni, @fabiocos, @davidlange6 can you please check and sign again.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7aac08/153/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-7aac08/11624.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2021_GenSimFull+DigiFull_2021+RecoFull_2021+HARVESTFull_2021+ALCAFull_2021

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3211964
 - DQMHistoTests: Total failures: 132
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211628
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

andresfelquintero commented:
Are there any other tests required to merge this into the master branch?

andresfelquintero commented:
Hello, any updates from this pull request? Are the commits appropriate?

fabiocos commented:
@andresfelquintero your PR in itself looks ok to me (at least for the prompt part, express to be added), but it is ineffective because of an apparent bug in ConfigBuilder.py . After
https://cmssdt.cern.ch/lxr/source/Configuration/Applications/python/ConfigBuilder.py#2229
you need to add:
```
            self.process.options.numberOfThreads=cms.untracked.uint32(int(self._options.nThreads))
            self.process.options.numberOfStreams=cms.untracked.uint32(int(self._options.nStreams))            
 self.process.options.numberOfConcurrentLuminosityBlocks=cms.untracked.uint32(int(self._options.nConcurrentLumis))
```
otherwise the option that you pass will never be propagated to the final configuration. I have tested that this works, without changing the behaviour of cmsDriver that does not seem to rely on process as built in prepare() but on the pythonCfgCode content (correctly set).

andresfelquintero committed:
Updated ConfigBuilder.py

I added the lines that Fabio suggested on https://github.com/cms-sw/cmssw/pull/26684

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26684/9826

- This PR adds an extra 48KB to repository

cmsbuild commented:
Pull request #26684 was updated. @cmsbuild, @franzoni, @fabiocos, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 75d6c6d2ba2b0119f6ffe1aabeaa494d5c2d8661
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7aac08/281/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/fc4b3671c67db7fa8a3cb98e48ca76946e31a915
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26684/281/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26684/281/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7aac08/281/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3212004
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211668
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

fabiocos commented:
@andresfelquintero as far as I am concerned this looks ok, unless @slava77 and @davidlange6 have further comment it could move forward. I think Express should be instrumented in the same way

fabiocos commented:
+operations

the code provides the desired functionality, fixing a bug in ConfigBuilder when used through DataProcessing (apparently ineffective so far)

fabiocos commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.
